platforms/core-configuration/declarative-dsl-evaluator/src/main/kotlin/org/gradle/internal/declarativedsl/evaluator/schema/InterpretationSchemaBuilder.kt

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.declarativedsl.evaluator.schema
+
+import org.gradle.declarative.dsl.evaluation.InterpretationSequence
+
+
+interface InterpretationSchemaBuilder {
+    fun getEvaluationSchemaForScript(
+        scriptContext: RestrictedScriptContext,
+    ): InterpretationSchemaBuildingResult
+}
+
+
+sealed interface InterpretationSchemaBuildingResult {
+    class InterpretationSequenceAvailable(val sequence: InterpretationSequence) : InterpretationSchemaBuildingResult
+    data object SchemaNotBuilt : InterpretationSchemaBuildingResult
+}
+
+
+sealed interface RestrictedScriptContext {
```

Comment:
💅 Shouldn't we rename this to something like `DeclarativeScriptContext`?

Reply:
Done in https://github.com/gradle/gradle/pull/29276/commits/c8cc3ef7a1132d012c0c5a63181cfb280149c095